lens/util/store.go

```diff
@@ -115,3 +127,186 @@ func (pb *ProxyingBlockstore) GetCount() int64 {
+}
+
+func NewCachingBlockstore(blocks blockstore.Blockstore, cacheSize int) (*CachingBlockstore, error) {
+	cache, err := lru.NewARC(cacheSize)
+	if err != nil {
+		return nil, xerrors.Errorf("new arc: %w", err)
+	}
+
+	return &CachingBlockstore{
+		cache:  cache,
+		blocks: blocks,
+	}, nil
+}
+
+func (cs *CachingBlockstore) DeleteBlock(c cid.Cid) error {
+	return cs.blocks.DeleteBlock(c)
+}
+
+func (cs *CachingBlockstore) GetSize(c cid.Cid) (int, error) {
+	return cs.blocks.GetSize(c)
+}
+
+func (cs *CachingBlockstore) Put(blk blocks.Block) error {
+	return cs.blocks.Put(blk)
+}
+
+func (cs *CachingBlockstore) PutMany(blks []blocks.Block) error {
+	return cs.blocks.PutMany(blks)
+}
+
+func (cs *CachingBlockstore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	return cs.blocks.AllKeysChan(ctx)
+}
+
+func (cs *CachingBlockstore) HashOnRead(enabled bool) {
+	cs.blocks.HashOnRead(enabled)
+}
+
+func (cs *CachingBlockstore) DeleteMany(cids []cid.Cid) error {
+	return cs.blocks.DeleteMany(cids)
+}
+
+func (cs *CachingBlockstore) Get(c cid.Cid) (blocks.Block, error) {
+	reads := atomic.AddInt64(&cs.reads, 1)
+	if reads%1000000 == 0 {
+		hits := atomic.LoadInt64(&cs.hits)
+		by := atomic.LoadInt64(&cs.bytes)
+		log.Debugw("CachingBlockstore stats", "reads", reads, "cache_len", cs.cache.Len(), "hit_rate", float64(hits)/float64(reads), "bytes_read", by)
+	}
+
+	v, hit := cs.cache.Get(c)
+	if hit {
+		atomic.AddInt64(&cs.hits, 1)
+		return v.(blocks.Block), nil
+	}
+
+	blk, err := cs.blocks.Get(c)
+	if err != nil {
+		return nil, err
+	}
+
+	atomic.AddInt64(&cs.bytes, int64(len(blk.RawData())))
+	cs.cache.Add(c, blk)
+	return blk, err
+}
+
+func (cs *CachingBlockstore) View(c cid.Cid, callback func([]byte) error) error {
+	reads := atomic.AddInt64(&cs.reads, 1)
+	if reads%1000000 == 0 {
+		hits := atomic.LoadInt64(&cs.hits)
+		by := atomic.LoadInt64(&cs.bytes)
+		log.Debugw("CachingBlockstore stats", "reads", reads, "cache_len", cs.cache.Len(), "hit_rate", float64(hits)/float64(reads), "bytes_read", by)
+	}
+	v, hit := cs.cache.Get(c)
+	if hit {
+		atomic.AddInt64(&cs.hits, 1)
+		return callback(v.(blocks.Block).RawData())
+	}
+
+	blk, err := cs.blocks.Get(c)
+	if err != nil {
+		return err
+	}
+
+	atomic.AddInt64(&cs.bytes, int64(len(blk.RawData())))
+	cs.cache.Add(c, blk)
+	return callback(blk.RawData())
+}
+
+func (cs *CachingBlockstore) Has(c cid.Cid) (bool, error) {
+	atomic.AddInt64(&cs.reads, 1)
+	// Safe to query cache since blockstore never deletes
+	if cs.cache.Contains(c) {
+		return true, nil
+	}
+
+	return cs.blocks.Has(c)
+}
+
+var _ adt.Store = (*CachingStateStore)(nil)
+
+type CachingStateStore struct {
+	cache  *lru.ARCCache
+	blocks blockstore.Blockstore
+	store  adt.Store
+	reads  int64 // updated atomically
+	hits   int64 // updated atomically
+}
+
+func NewCachingStateStore(blocks blockstore.Blockstore, cacheSize int) (*CachingStateStore, error) {
+	cache, err := lru.NewARC(cacheSize)
```

Comment:
iirc this only errors when a negative value is passed, could panic here instead and simplify some of the calling code.

Reply:
I don't think it's good UX to panic on user input

Reply:
True, and enforcement of positive integers could be done with `Uint` flags, but I'm alright with this as is.